QA-131: Add a .gitlab-ci file which can be included in other files

oleorhagen opened the pull request:
This commit adds a '.gitlab-ci-black-format-code.yml' file which can be included
into other projects' .gitlab-ci file, and which, when included, will check that
the Python3 code in the given directories are formatted according to Black's
preferences.

Changelog: Title

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

lluiscampos commented on .gitlab-ci-black-format-code.yml:
```diff
@@ -0,0 +1,73 @@
+# .gitlab-ci-black-format-code.yml
```
I would explicitly mention pytrhon3 in the filename.

The reason is that we could also have one for the go format (for which we have duplicated code all around the pipelines).

lluiscampos commented on .gitlab-ci-black-format-code.yml:
```diff
@@ -0,0 +1,73 @@
+#
+#[tool.black]
+# target-version = ['py38']
+# include = '\.pyi?$' # Regular expression
```
What is the `i` for? I nevern seen a .pyi file.

oleorhagen replied:
It was copy-pasted from the Black documentation. It is for `python-stubs` whatever that is. Removing.

lluiscampos commented on .gitlab-ci-black-format-code.yml:
```diff
@@ -0,0 +1,73 @@
+#                      # the root of the project
+# )
+# '''
+# The tools also ignores all files found in .gitignore.
```
But not `.git` folder?

oleorhagen replied:
Nope, it does seem to keep away from the git folder automatically.

lluiscampos commented on .gitlab-ci-black-format-code.yml:
```diff
@@ -0,0 +1,73 @@
+# )
+# '''
+# The tools also ignores all files found in .gitignore.
+
```
Really nice documentation! Thanks for setting an standard here :clap:

lluiscampos commented on .gitlab-ci-black-format-code.yml:
```diff
@@ -0,0 +1,73 @@
+    # Install dependencies
+    - apk add --no-cache git bash python3
+    - pip3 install --upgrade pip setuptools
+    - pip3 install black # TODO -- Should we freeze the version here (?)
```
I think so.

oleorhagen replied:
Alright, freezing!

lluiscampos commented on .gitlab-ci-black-format-code.yml:
```diff
@@ -0,0 +1,73 @@
+    - git fetch --tags origin
+    - git clone http://github.com/mendersoftware/mendertesting /tmp/mendertesting
+  script:
+    # Check that the python code is formattet
```
Python, with capital P, please.

lluiscampos commented on .gitlab-ci-black-format-code.yml:
```diff
@@ -0,0 +1,73 @@
+  script:
+    # Check that the python code is formattet
+    # If no arguments are provided, the script will check the current directory.
+    - /tmp/mendertesting/check_python_format "$CHECK_PYTHON_FORMAT_ARGS"
```
These arguments are not documented anywhere. This is something that the caller pipelines can use, so I suggest we add it to the above documentation.

lluiscampos commented on check_python_format:
```diff
@@ -0,0 +1,96 @@
+#
+# A simple script for verifying that all the commits in the PR-range has
+# the code correctly formatted with the Black code formatter.
+#
```
Is this true? By reading the script, I think it will verify the given directory (`.` by default) regardless of the PR-range.

lluiscampos replied:
I think we should check only the PR-range, as you have intended in the docs, so that we don't force huge PRs just formatting (which I personally dislike as they kill `git blame`)

lluiscampos commented on check_python_format:
```diff
@@ -0,0 +1,96 @@
+            echo "-d|--directory"
+            echo "  Adds the directory to the directories checked by Black"
+            echo ""
+            echo "This script runs the black formatter on all the python code"
```
Python

lluiscampos commented on check_python_format:
```diff
@@ -0,0 +1,96 @@
+    echo "Checking range: ${COMMIT_RANGE}:"
+    git --no-pager log $COMMIT_RANGE
+    commits="$(git rev-list --no-merges $COMMIT_RANGE)"
+fi
```
Up until here the logic is the same as the check commits script. I wonder if we could move it to a common script with a function that then both could source... ;)

oleorhagen replied:
he he, yup. It is in the pipes ;)

lluiscampos commented on check_python_format:
```diff
@@ -0,0 +1,96 @@
+            echo >&2 "$COMMIT_MSG"
+            echo >&2 "----------------------------------------"
+            echo >&2 "Culprit: $COMMIT_USER_EMAIL"
+            exit 1
```
As I commented at the start, I don't see the relation between `$commits` and `$CODE_DIRECTORIES`. The full messaging here seems bogus to me.

oleorhagen commented:
@lluiscampos This brings it closer to what we want, I think.

It now filters on directories, and is only run on python files which are part of a commit.

The filter is specified through an environment variable `CODE_DIRECTORIES`, which will make sure, only files from this directory will be run through the formatter.

Note that this implies that all code should be base formatted, before enabling this functionality.

oleorhagen commented:
Sample output when run on `master`:
```
Checking range: HEAD~1..HEAD:
commit 872e18ed8602696b338048eb1845b3339862b707
Merge: ecce811 dd5c44e
Author: Kristian Amlie <kacf@users.noreply.github.com>
Date:   Wed Dec 11 16:14:32 2019 +0100

    Merge pull request #876 from oleorhagen/MEN-2956
    
    Update the acceptance tests to handle default rootfs-image-checksum provides

commit dd5c44e67b55870e24298898bf3ef0981b02680b
Author: Ole Petter <ole.orhagen@northern.tech>
Date:   Tue Dec 10 11:25:10 2019 +0100

    MEN-2956: acceptance/ Handle default rootfs-image-checksum provides
    
    Update the test_build_depends_and_provides acceptance tests to handle the
    default `rootfs_image_checksum` now added by `mender-artifact write
    rootfs-image`.
    
    Changelog: None
    
    Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

commit 63efe559a0edfa9a406260d7019088910aa9b8c4
Author: Ole Petter <ole.orhagen@northern.tech>
Date:   Tue Dec 10 10:17:39 2019 +0100

    Housecleaning: whitespace cleanup
    
    Changelog: None
    
    Signed-off-by: Ole Petter <ole.orhagen@northern.tech>

Verifying that the Python code in dd5c44e67b55870e24298898bf3ef0981b02680b
and directory ./ is formatted correctly...
Using configuration from /home/olepor/go/src/github.com/mendersoftware/meta-mender/pyproject.toml.
would reformat tests/acceptance/test_build.py
All done! 💥 💔 💥
1 file would be reformatted.
The Python code in commit dd5c44e67b55870e24298898bf3ef0981b02680b
is not formatted correctly, according to Black.
Please format the code correctly, and commit anew.

Commit:
--------------------------------------------------------------------------------
MEN-2956: acceptance/ Handle default rootfs-image-checksum provides

Update the test_build_depends_and_provides acceptance tests to handle the
default `rootfs_image_checksum` now added by `mender-artifact write
rootfs-image`.

Changelog: None

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>
--------------------------------------------------------------------------------
Culprit(s): 
tests/acceptance/test_build.py

Verifying that the Python code in 63efe559a0edfa9a406260d7019088910aa9b8c4
and directory ./ is formatted correctly...
Using configuration from /home/olepor/go/src/github.com/mendersoftware/meta-mender/pyproject.toml.
would reformat tests/acceptance/test_build.py
All done! 💥 💔 💥
1 file would be reformatted.
The Python code in commit 63efe559a0edfa9a406260d7019088910aa9b8c4
is not formatted correctly, according to Black.
Please format the code correctly, and commit anew.

Commit:
--------------------------------------------------------------------------------
Housecleaning: whitespace cleanup

Changelog: None

Signed-off-by: Ole Petter <ole.orhagen@northern.tech>
--------------------------------------------------------------------------------
Culprit(s): 
tests/acceptance/test_build.py
```

oleorhagen commented:
@lluiscampos Ready for review.

lluiscampos commented on check_python_code_format:
```diff
@@ -0,0 +1,131 @@
+        echo >&2 "--------------------------------------------------------------------------------"
+        echo >&2 "$COMMIT_MSG"
+        echo >&2 "--------------------------------------------------------------------------------"
+        printf >&2 "Culprit(s): $FAILED_FILES\n"
```
After this, I would help the user fixing it's mess. For that, I would add something like:

`Please fix your code and check locally with: black --check --verbose  "$file"`

(I think we can assume that he/she knows know to install `black`)

oleorhagen commented on .gitlab-ci-format-python3-code.yml:
```diff
@@ -0,0 +1,60 @@
+# 
+# include:
+#   - project: 'Northern.tech/Mender/mendertesting'
+#     file: '.gitlab-ci-black-format-python3-code.yml'
```
I like it!
